### ISSUE TITLE
Random crashes with Kirby Super Star Ultra

### ISSUE THREAD

**Issue by FarLanderCraft** [Arisotura/melonDS#1156]
System: Arch Linux x86_64
Kernel: 5.12.15-zen1
DE: Plasma 5.22.3
melonDS version: 0.9.2

Issue: Random crashes with Kirby Super Star Ultra

Steps to recreate: Get to a point to where it's either going from sprite graphics to cutscene, or from cutscene to sprite graphics. It works most of the time, but will occasionally crash. If you make a save state before the transition, and it crashes after where you made the save state, when you load the save state it will usually progress like normal.

Notes: This was not an issue in previous versions, though I am unsure which version introduced the issue. However I can say with confidence that Kirby Super Star Ultra didn't randomly crash in version 0.8.3.

**Comment by RSDuck** [comment]
can you give us a savefile for fast reproduction? Also do you have the JIT recompiler or the OpenGL renderer enabled?

**Comment by FarLanderCraft** [comment]
JIT recompiler is enabled, OpenGL renderer isn't.
As for the save file, here you go.
[Kirby Save (GitHub wanted zip).zip](https://github.com/Arisotura/melonDS/files/6817167/Kirby.Save.GitHub.wanted.zip.zip)

**Comment by FarLanderCraft** [comment]
(Don't know if it helps, but it's the latest version in the AUR. Not sure if it's officially maintained or not. If needbe I can build from master, and test with that instead.)

**Comment by RSDuck** [comment]
yes please test with master as well.

**Comment by FarLanderCraft** [comment]
Seems to not occur at all on master, I wonder how outdated the AUR build is.
Well, sorry for the trouble lol .

**Comment by FarLanderCraft** [comment]
Never mind it's still happening, I just got lucky.

**Comment by FarLanderCraft** [comment]
I was streaming Spring Breeze on Discord, and it crashed at the beginning of the end cutscene when King Dedede gets launched out of the castle. ;-;

**Comment by RSDuck** [comment]
can you try playing with JIT disabled and tell me whether this makes a difference?

I've been trying to reproduce it, but I wasn't successful yet. Is there any pattern or game in particular where this happened?

**Comment by RSDuck** [comment]
also have you ever used gdb?

**Comment by FarLanderCraft** [comment]
> also have you ever used gdb?
> 
> EDIT: also what exactly do you mean by crash? Does the emulator close or does the game just freeze?

Emulator crashes. I don't think I've used gdb. And I can't see any pattern. It doesn't always happen, it seems pretty rare on master. (Happened much more often on the AUR build.)

**Comment by FarLanderCraft** [comment]
As for playing with JIT disabled, I'll have to get back to you on that one, because it might take a while to reach a conclusion one way or another.

**Comment by FarLanderCraft** [comment]
Okay never mind it crashed first try with JIT disabled.

**Comment by RSDuck** [comment]
ok thanks for the information.

Can you please try getting information on where the crash happens?

To do this you need to do the following:
Make a debug build (by adding `-DCMAKE_BUILD_TYPE=Debug` to the cmake flags then recompile). Then start melonDS by running `gdb ./melonDS`. Once there's a prompt enter `start`. It might stop again and saying temporary break point or so, then enter `continue` to well continue. Make sure the JIT recompiler is disabled and then try to get it crashing. When this happens it should very likely say something about segfault and some other information. There there should be a prompt again. Enter the command `backtrace` there. Copy everything printed during this process and paste it here.

Something which could happen is that the bug doesn't happen in debug builds, which would make this whole endeavour quite nasty, but hopefully it crashes in debug mode as well.

**Comment by FarLanderCraft** [comment]
Would it be alright if I use a save state I make before an animation and load state until it happens again or do I have to replay a part until it happens?

**Comment by RSDuck** [comment]
if you know that this can reproduces the same crash, sure.

**Comment by FarLanderCraft** [comment]
I haven't been able to get it to crash again yet, but something I've noticed is the console gets spammed with `NDS SRAM: Flush requested` every transition (where the chance for crashing is), could that have something to do with it?
![image](https://user-images.githubusercontent.com/67613281/125695059-f56ffc29-9eaa-4f0a-936f-55f2f5ee5d24.png)

**Comment by RSDuck** [comment]
nah that's completely normal, it just means that the game is saving to cartridge save memory.

**Comment by FarLanderCraft** [comment]
Yeah it seems to not crash in Debug mode.
I'll keep trying a little longer, but I'm getting tired of pressing F1. ;-;

**Comment by FarLanderCraft** [comment]
> nah that's completely normal, it just means that the game is saving to cartridge save memory.

Is that why the save file corrupted when it crashed?

**Comment by RSDuck** [comment]
yes, it could be possible that the crash happened while it hadn't saved everything yet, thus rendering the saved state invalid.

Was it previously possible to get the crash from loading a savestate?

**Comment by FarLanderCraft** [comment]
After a bit I used save states to recover. Like, I'd make a save state before every level transition or cutscene transition.
If it crashed, I'd restart melonDS, load the save state, then it'd usually work from there without crashing.
I think it might be some form of instability.

**Comment by FarLanderCraft** [comment]
Also it seems you misunderstood what I maent by save file.
It wasn't a save state, it was the actual game save.

**Comment by FarLanderCraft** [comment]
If it helps my CPU is a Ryzen 7 1700X, my GPU is a Radeon RX 570, I have 16GB DDR4-3000 (Dual-channel) RAM, the KSSU ROM is on a 2TB 7200 RPM hard disk, however loading from my PCIe x2 SSD also had the occasional crash. I don't know how much of this is relevant, or if any of this is necessary information, but I am usually asked about my hardware config at this point when I seek support on Discord.

**Comment by RSDuck** [comment]
hw shouldn't matter for bugs like this.

**Comment by FarLanderCraft** [comment]
After loading Pokemon Diamond I'm having the same problems, it is related to save files, I will now try it in debug mode.

**Comment by RSDuck** [comment]
can you come to our IRC to discuss this better?

**Comment by FarLanderCraft** [comment]
I've never used an IRC client lol.
I mean I can if you'd really prefer it.

**Comment by RSDuck** [comment]
please writing on github is a bit exhausting

**Comment by FarLanderCraft** [comment]
I can understand that, alright.
